### PR TITLE
pkg/generator: Create generic file rendering function

### PR DESCRIPTION
**Pull request by theishshah**
Issue #303

### CODE REVIEW

**Comment by theishshah** [comment]
This had been tested against the memcached user guide example and _appears_ to be in working order

**Comment by theishshah** [comment]
So generator.go still needs to be cleaned up, but this offers a single render function instead of the 9 separate files and functions which did functionally the same work of loading a template and replacing values from a struct

**Comment by theishshah** [comment]
Fixed issue with creating olm directory

File: pkg/generator/generator.go [hunk 1]
```diff
@@ -351,52 +415,159 @@ func (g *Generator) renderPkg() error {
...
+	}
+	sDir := filepath.Join(g.projectName, stubDir)
+	if err := os.MkdirAll(sDir, defaultDirFileMode); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func renderGenericFile(w io.Writer, fileLoc string, fileTmpl string, info tmplData) error {
```
Comment (rithujohn191) [inline pkg/generator/generator.go]: Could you rename this to `renderFile` and add a comment above it describing what it does?

File: pkg/generator/generator.go [hunk 2]
```diff
@@ -351,52 +415,159 @@ func (g *Generator) renderPkg() error {
...
+	PackageName string
+	ChannelName string
+	CurrentCSV  string
+
+	CRDVersion     string
+	CSVName        string
+	CatalogVersion string
+}
+
+func (g *Generator) generateDirStructure() error {
```
Comment (rithujohn191) [inline pkg/generator/generator.go]: Could you please add a small comment describing what the function does.

File: pkg/generator/generator.go [hunk 3]
```diff
@@ -351,52 +415,159 @@ func (g *Generator) renderPkg() error {
...
+		Kind:              kind,
+		APIDirName:        apiDirName,
+		Version:           version,
+	}); err != nil {
 		return err
 	}
 	return writeFileAndPrint(filepath.Join(stubDir, handler), buf.Bytes(), defaultFileMode)
 }
 
+type tmplData struct {
```
Comment (rithujohn191) [inline pkg/generator/generator.go]: Most of these fields are pretty self-explanatory. But for fields like `KindSingular` and `KindPlural` could you add a comment please

Comment (theishshah) [reply pkg/generator/generator.go]: @hasbro17 I wasn't here when the interfaces were designed, can you provide some insight as to why we have 3 versions of Kind?

Comment (theishshah) [reply pkg/generator/generator.go]: Cool, just updated, should be ready to merge after Travis finishes

File: pkg/generator/generator.go [hunk 4]
```diff
@@ -157,25 +172,28 @@ func (g *Generator) renderGoDep() error {
...
-		return err
-	}
 	return renderCmdFiles(cpDir, g.repoPath, g.apiVersion, g.kind)
 }
 
 func renderCmdFiles(cmdProjectDir, repoPath, apiVersion, kind string) error {
 	buf := &bytes.Buffer{}
-	if err := renderMainFile(buf, repoPath, apiVersion, kind); err != nil {
+
+	if err := renderFile(buf, "cmd/<projectName>/main.go", mainTmpl, tmplData{
```
Comment (hasbro17) [inline pkg/generator/generator.go]: Can you define the tmplData argument before the function call. It helps to improve readability:
```Go
	td := tmplData{
		OperatorSDKImport: sdkImport,
		StubImport:        filepath.Join(repoPath, stubDir),
		K8sutilImport:     k8sutilImport,
		SDKVersionImport:  versionImport,
		APIVersion:        apiVersion,
		Kind:              kind,
	}
	if err := renderFile(buf, "cmd/<projectName>/main.go", mainTmpl, td); err != nil {
		return err
	}
```

Comment (hasbro17) [reply pkg/generator/generator.go]: Here and in all other places.

File: pkg/generator/generator.go [hunk 5]
```diff
@@ -351,52 +435,167 @@ func (g *Generator) renderPkg() error {
...
+	PackageName string
+	ChannelName string
+	CurrentCSV  string
+
+	CRDVersion     string
+	CSVName        string
+	CatalogVersion string
+}
+
+//Creates all the necesary directories for the generated files
```
Comment (hasbro17) [inline pkg/generator/generator.go]: Missing space in comment: `//Creates` ==> `// Creates`

File: pkg/generator/generator.go [hunk 6]
```diff
@@ -351,52 +435,167 @@ func (g *Generator) renderPkg() error {
...
+	}
+	sDir := filepath.Join(g.projectName, stubDir)
+	if err := os.MkdirAll(sDir, defaultDirFileMode); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//Renders a file given a template, and fills in template fields according to values passed in the tmplData struct
```
Comment (hasbro17) [inline pkg/generator/generator.go]: Missing space in comment: `//Renders ` ==> `// Renders `

**Pull request merged**